### PR TITLE
BTFS-663: new metadata implementation

### PR DESCRIPTION
**Pull request by steveyeom**
New metadata implementation:
1. Several unit tests including reading metadata plus user data are created and used to verify the 
  required functionality: 
   - importer/balanced/balanced_test.go
2. Modified importer/balanced/builder.go to create a new Metadata building driver, BuildMetadataDag(..), and its supporting functions.
3. Created a new file "importer/helpers/metaDagBuilder.go" for MetadataDagBuilderHelper to extend existing DagBuilderHelper.
4. Modified "importer/helpers/dagBuilder.go" to create supporting getters and setters.

### CODE REVIEW

File: importer/balanced/balanced_test.go [hunk 1]
```diff
@@ -337,3 +337,152 @@ func TestSeekingConsistency(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func buildMetaTestDag(ds ipld.DAGService, maxlinks int, dataspl chunker.Splitter, metadata []byte, chunksize int64) (*dag.ProtoNode, error) {
+	dbp := h.DagBuilderParams{
+		Dagserv:       ds,
+		Maxlinks:      maxlinks,
+		TokenMetadata: metadata,
+		ChunkSize:     uint32(chunksize),
```
Comment (taiyangc) [inline importer/balanced/balanced_test.go]: uint64 now

File: importer/balanced/balanced_test.go [hunk 2]
```diff
@@ -337,3 +337,152 @@ func TestSeekingConsistency(t *testing.T) {
...
+	b := []byte(`{"nodeid":"QmURnhjU6b2Si4rqwfpD4FDGTzJH3hGRAWSQmXtagywwdz","Price":12.4}`)
+	testMetaDataConsistency(t, 4, b, 8)
+}
+
+func buildTestDagWithMetadata(ds ipld.DAGService, maxlinks int, dataspl chunker.Splitter, metadata []byte, chunksize int64) (*dag.ProtoNode, error) {
+	dbp := h.DagBuilderParams{
+		Dagserv:       ds,
+		Maxlinks:      maxlinks,
+		TokenMetadata: metadata,
+		ChunkSize:     uint32(chunksize),
```
Comment (taiyangc) [inline importer/balanced/balanced_test.go]: uint64

File: importer/balanced/balanced_test.go [hunk 3]
```diff
@@ -337,3 +337,152 @@ func TestSeekingConsistency(t *testing.T) {
...
+	}
+
+	return nd, data, metadata
+}
+
+func getRootsForDataAndMetadata(t *testing.T, root ipld.Node, ds ipld.DAGService) (*dag.ProtoNode, *dag.ProtoNode, error) {
+	var nodes [2]ipld.Node
+
+	for i := 0; i < 2; i++ {
+		lnk := root.Links()[i]
```
Comment (taiyangc) [inline importer/balanced/balanced_test.go:455]: should check if `len(root.Links()) >= 2` first

File: importer/balanced/balanced_test.go [hunk 4]
```diff
@@ -337,3 +337,152 @@ func TestSeekingConsistency(t *testing.T) {
...
+		nodes[i] = child
+	}
+	return nodes[0].(*dag.ProtoNode), nodes[1].(*dag.ProtoNode), nil
+}
+
+func testTokenMetadataRead(t *testing.T, dsize int64, maxlinks int, mdata []byte, chksize int64) {
+	ds := mdtest.Mock()
+	nd, should, mshould := getTestDagWithMetadata(t, ds, dsize, maxlinks, mdata, chksize)
+
+	mnd, dnd, err := getRootsForDataAndMetadata(t, nd, ds)
```
Comment (taiyangc) [inline importer/balanced/balanced_test.go:470]: check  err here

File: importer/balanced/builder.go [hunk 5]
```diff
@@ -131,7 +132,7 @@ import (
 //
 func Layout(db *h.DagBuilderHelper) (ipld.Node, error) {
 	if !db.IsMultiDagBuilder() {
-		return layout(db)
+		return layout(db, db.IsThereMetaData() && !db.IsReedSolomon())
```
Comment (taiyangc) [inline importer/balanced/builder.go]: Why is it not supporting reed solomon + metadata?

File: importer/balanced/builder.go [hunk 6]
```diff
@@ -211,23 +213,67 @@ func layout(db *h.DagBuilderHelper) (ipld.Node, error) {
...
+
+	// Add token metadata DAG, if exists, as a child to the 'newRoot'.
+	if addMetaDag {
+		root, err = attachMetadataDag(db, root, fileSize)
 		if err != nil {
 			return nil, err
 		}
 	}
 
+	if DEBUG {
```
Comment (taiyangc) [inline importer/balanced/builder.go]: Not a very common way to debug, if you really need it, consider using something like: https://github.com/TRON-US/go-btfs/blob/master/cmd/btfs/main.go#L195 which is environment based.

File: importer/balanced/builder.go (hunk 5 above)
Comment (taiyangc) [reply importer/balanced/builder.go]: Also, dagbuilder should not understand reed solomon. All it understands is that it is a "multi dag builder". So it should not explicitly set a reed solomon flag. Instead, use `db.IsMultiDagBuilder` to check if the reed solomon-like dag is used.

File: importer/balanced/builder.go [hunk 7]
```diff
@@ -318,3 +363,49 @@ func fillNodeRec(db *h.DagBuilderHelper, node *h.FSNodeOverDag, depth int) (fill
 
 	return filledNode, nodeFileSize, nil
 }
+
+func attachMetadataDag(db *h.DagBuilderHelper, root ipld.Node, fileSize uint64) (ipld.Node, error) {
+	// Create a 'newRoot'.
+	newRoot := db.NewFSNodeOverDag(ft.TFile)
+
+	// Add metadata DAG as first child of 'newRoot'.
+	err := addMetadataChild(db, newRoot, db.GetMetaDb().GetMetaDagRoot())
```
Comment (taiyangc) [inline importer/balanced/builder.go]: So, can you confirm this: metadata (if available) is always the left most child.

File: importer/helpers/dagbuilder.go [hunk 8]
```diff
@@ -12,27 +13,33 @@ import (
...
+	spl        chunker.Splitter
+	recvdErr   error
+	rawLeaves  bool
+	nextData   []byte // the next item to return.
+	maxlinks   int
+	cidBuilder cid.Builder
+
+	metaDb       *MetaDagBuilderHelper
+	metaDagBuilt bool
+	reedSolomon  bool
```
Comment (taiyangc) [inline importer/helpers/dagbuilder.go]: See comment above, this flag is not necessary, use `IsMultiDagBuilder()` to check for reed solomon case.

File: importer/helpers/dagbuilder.go [hunk 9]
```diff
@@ -175,22 +211,22 @@ func (db *DagBuilderHelper) Next() ([]byte, error) {
 
 // GetDagServ returns the dagservice object this Helper is using
 func (db *DagBuilderHelper) GetDagServ() ipld.DAGService {
-	return db.dserv
+	return db.Dserv
```
Comment (taiyangc) [inline importer/helpers/dagbuilder.go]: Why expose `Dserv`? You can just use `GetDagServ()` everywhere as is.

File: importer/helpers/dagbuilder.go [hunk 10]
```diff
@@ -175,22 +211,22 @@ func (db *DagBuilderHelper) Next() ([]byte, error) {
...
 
 // GetCidBuilder returns the internal `cid.CidBuilder` set in the builder.
 func (db *DagBuilderHelper) GetCidBuilder() cid.Builder {
 	return db.cidBuilder
 }
 
-// GetTokenMetadata returns token metadata []byte.
-func (db *DagBuilderHelper) GetTokenMetadata() []byte {
-	return db.tokenMetadata
+// isThereMetaData returns whether token metadata exists.
```
Comment (taiyangc) [inline importer/helpers/dagbuilder.go:224]: `IsThereMetadata`

File: importer/helpers/dagbuilder.go [hunk 11]
```diff
@@ -403,6 +427,18 @@ func (n *FSNodeOverDag) AddChild(child ipld.Node, fileSize uint64, db *DagBuilde
 	return db.Add(child)
 }
 
+// AddChildDag adds the DAG topped by the given "child"
+// whose DAG descendents are already added into blockservice.
+func (n *FSNodeOverDag) AddChildDag(child ipld.Node, fileSize uint64, db *DagBuilderHelper) error {
+	err := n.dag.AddNodeLink("", child)
```
Comment (taiyangc) [inline importer/helpers/dagbuilder.go:435]: Can't you just use `AddChild` everywhere? Why is there a need for original `AddChildDag`. `DagBuilderHelperInterface` is the generic form.

File: importer/helpers/metaDagBuilder.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
+package helpers
```
Comment (taiyangc) [inline importer/helpers/metaDagBuilder.go]: follow golang file name convention, use `metadagbuilder.go` or `meta_dagbuilder.go`

File: importer/helpers/metaDagBuilder.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+func (ms *MetaSplitter) Reader() io.Reader {
+	return ms.r
+}
+
+func NewMetaSplitter(r io.Reader, size int64) chunker.Splitter {
+	return &MetaSplitter{
+		r:    r,
+		size: uint64(size),
+	}
+}
```
Comment (taiyangc) [inline importer/helpers/metaDagBuilder.go]: This looks exactly like `SizeSplitter`, why another type? Even if you need it, the change needs to be in `go-btfs-chunker`.

File: importer/helpers/metaDagBuilder.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+type DagBuilderHelperInterface interface {
+	NewFSNodeOverDag(pb.Data_DataType) *FSNodeOverDag
+	Maxlinks() int
+	Done() bool
+	Add(ipld.Node) error
+	NewLeafDataNode(pb.Data_DataType) (ipld.Node, uint64, error)
+}
```
Comment (taiyangc) [inline importer/helpers/metaDagBuilder.go]: The interface should go into the original `dagbuilder.go`

File: importer/helpers/metaDagBuilder.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	dataSize = uint64(len(metaDataChunk))
+
+	// Create a new leaf node with token metadata.
+	node, err = mdb.NewLeafNode(metaDataChunk, ft.TTokenMeta)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return node, dataSize, nil
+}
```
Comment (taiyangc) [inline importer/helpers/metaDagBuilder.go]: Seems to have a lot of similarities with the original dag builder ... which part is the exact code that needs modification? Let's chat.

File: importer/balanced/builder.go (hunk 5 above)
Comment (steveyeom) [reply importer/balanced/builder.go]: My assumption is when an individual splitter instance is running, db.IsMultiDagBuilder() will return false. But this flag will go away after BTFS-931.

File: importer/helpers/metaDagBuilder.go (hunk 12 above)
Comment (steveyeom) [reply importer/helpers/metaDagBuilder.go]: done

File: importer/helpers/metaDagBuilder.go (hunk 14 above)
Comment (steveyeom) [reply importer/helpers/metaDagBuilder.go]: done

File: importer/helpers/dagbuilder.go (hunk 11 above)
Comment (steveyeom) [reply importer/helpers/dagbuilder.go:435]: The code of AddChildDag() is different than AddChild().

**Comment by steveyeom** [comment]
The last changes include 
1. Minor changes 
2. Changes to make all the unit tests in go-unixfs pass 
3. Changes to make reed-solomon work including one new unit test

File: importer/balanced/builder.go (hunk 5 above)
Comment (taiyangc) [reply importer/balanced/builder.go]: Got it.

File: importer/helpers/dagbuilder.go (hunk 11 above)
Comment (taiyangc) [reply importer/helpers/dagbuilder.go:435]: ok.

File: importer/helpers/dagbuilder.go (hunk 9 above)
Comment (taiyangc) [reply importer/helpers/dagbuilder.go]: How about this question?

**Comment by steveyeom** [comment]
Again all the unit tests inside go-unixfs passed w/o failure with the changes.

**Review by taiyangc — APPROVED** [review]

**Pull request merged**